concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -479,88 +481,99 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
+      ur <- case tr of
+          WithMetadata{wmdData = NormalTransaction tx} -> do
+              -- Don't accept the transaction if the stated energy is smaller than the minimum energy amount.
+              let baseEnergy = baseCost (getTransactionHeaderPayloadSize $ transactionHeader tx) (getTransactionNumSigs $ transactionSignature tx)
+                  statedEnergy = thEnergyAmount $ transactionHeader tx
+              if baseEnergy > statedEnergy then return ResultTooLowEnergy
+              else do
+                 -- this is not ideal since we look up the entire account.
+                 -- In the current implementation this is not a problem since all states since the last finalized one are cached
+                 -- but this should be revised to add a "accountExists" function in the future
+                 senderExists <- flip getAccount (transactionSender tx) =<< queryBlockState =<< lastFinalizedBlock
+                 if isNothing senderExists then return ResultNonexistingSenderAccount
+                 else snd <$> doReceiveTransactionInternal tr slot
+          _ -> snd <$> doReceiveTransactionInternal tr slot
+      when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
+      return ur
 
 -- |Add a transaction to the transaction table.  The 'Slot' should be
 -- the slot number of the block that the transaction was received with.
 -- This function should only be called when a transaction is received as part of a block.
 -- The difference from the above function is that this function returns an already existing
--- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
+-- transaction ipn case of a duplicate, ensuring more sharing of transaction data.
+-- This function also verifies the transactions incoming and adds them to the internal
+-- transaction verification cache such that it can be used by the 'Scheduler'.
+doReceiveTransactionInternal :: (TreeStateMonad pv m, SkovQueryMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
+doReceiveTransactionInternal tr slot = do
+  -- Don't accept the transaction if its expiry time is too far in the future
+   slotTime <- getSlotTime slot
+   expiryTooLate <- isExpiryTooLate tr slotTime
+   if expiryTooLate then return (Nothing,  ResultExpiryTooLate)
+   -- Don't accept the transaction if it was expired
+   else if transactionExpired (msgExpiry tr) (utcTimeToTimestamp slotTime) then return (Nothing, ResultStale)
+   else do
+    cache <- getTransactionVerificationCache
+    ts <- getSlotTimestamp slot
+    (verRes, cache') <- runReaderT (TV.verifyWithCache ts tr cache) =<< blockState =<< getFocusBlock
+    if not (TV.isVerifiable verRes) then do return (Nothing, mapTransactionVerificationResult verRes)
```

Comment:
Isn't this already done in `doReceiveTransaction`?

Reply:
Yes, I will rework this. 
The reason for the _duplicate_ check was that I wanted to make sure always to do the _cheap_ checks before e.g. signature checking. 
But I'll consolidate this into `doReceiveTransactionInternal`